lang/values.md

```diff
@@ -165,21 +165,21 @@ impl Type {
 
 ```rust
 /// Check if the given pointer is valid for safe pointer types (`Ref`, `Box`).
-fn check_safe_ptr(ptr: Pointer, pointee: Layout) -> bool {
+fn check_safe_ptr<M: Memory>(ptr: M::Pointer, pointee: Layout) -> bool {
     // References (and `Box`) need to be non-null, aligned, and not point to an uninhabited type.
     // (Think: uninhabited types have impossible alignment.)
     ptr.addr != 0 && ptr.addr % pointee.align.bytes() == 0 && pointee.inhabited
 }
 
 impl Type {
-    fn decode(Type::Ref { pointee, .. } | Type::Box { pointee }: Self, bytes: List<AbstractByte>) -> Option<Value> {
-        let ptr = decode_ptr(bytes)?;
-        if !check_safe_ptr(ptr, pointee) { throw!(); }
+    fn decode<M: Memory>(Type::Ref { pointee, .. } | Type::Box { pointee }: Self, bytes: List<M::AbstractByte>) -> Option<Value<M>> {
```

Comment:
Oh I forgot to update this for the unified pointer type. Will do that after your MR.
```suggestion
    // FIXME: this is not yet updated to the unified pointer type.
    fn decode<M: Memory>(Type::Ref { pointee, .. } | Type::Box { pointee }: Self, bytes: List<M::AbstractByte>) -> Option<Value<M>> {
```